applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -311,20 +311,34 @@ static int register_peripheral(struct bt_gatt_dm *dm, const uint8_t *hwid,
 	}
 	__ASSERT_NO_MSG(sub_id < ARRAY_SIZE(subscribers));
 
+	size_t per_free = ARRAY_SIZE(peripherals);
 	size_t per_id;
 	for (per_id = 0; per_id < ARRAY_SIZE(peripherals); per_id++) {
 		if (!is_peripheral_connected(&peripherals[per_id])) {
-			break;
+			per_free = per_id;
+			if (peripherals[per_id].sub_id == sub_id) {
+				break;
+			}
 		}
 	}
+	if (per_id == ARRAY_SIZE(peripherals)) {
+		per_id = per_free;
+	}
 	__ASSERT_NO_MSG(per_id < ARRAY_SIZE(peripherals));
 
-	peripherals[per_id].sub_id = sub_id;
+	struct hids_peripheral *per = &peripherals[per_id];
+
+	if (per->sub_id != sub_id) {
+		for (size_t report_id = 0; report_id < ARRAY_SIZE(per->enqueued_report); report_id++) {
+			drop_enqueued_reports(&per->enqueued_report[report_id]);
```

Comment:
We may drop a key release here. That would result in getting stuck with a pressed key on one of the subscribers.

Reply:
There is nothing that can be done with it. You cannot guarantee that all reports will be sent out.

Reply:
In case the USB subscriber does not unsubscribe/suspend, we know that all the enqueued reports will be actually sent to the subscriber.

Reply:
We may consider implementing HID report queue per subscriber (and remove queue per peripheral). For the single subscriber it would not be a problem since reports from the peripherals have different report IDs (if not then we may encounter many other problems anyway).

Reply:
As discussed off-line this could not be done without introducing problems with memory usage. I have however reworked the code so when peripheral is connected back it will get the unsent reports from the subscriber.